libs/UGridNET/dll/src/UGridReader.cs

```diff
@@ -0,0 +1,208 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UGridNET.Extensions;
+using System.Runtime.InteropServices;
+
+namespace UGridNET
+{
+
+    public sealed class UGridReader : UGridBase
+    {
+        // requires using log4net;
+        //private static readonly ILog log = LogManager.GetLogger(typeof(UGridReader));
+
+        private bool disposed = false;
+
+        public UGridReader(string path) : base(path, 0)
+        {
+            try
+            {
+                ReadMesh1D();
+                ReadMesh2D();
+                ReadContacts();
+                ReadNetwork1D();
+            }
+            catch
+            {
+                // If initialization fails, dispose of any partially initialized resources
+                Dispose();
+                // Rethrow to inform the caller of the failure
+                throw;
+            }
+        }
+
+        ~UGridReader()
+        {
+            // Cleanup in case Dispose wasn't called
+            Dispose(false);
+        }
+
+        public new void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+            base.Dispose();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposed)
+            {
+                if (disposing)
+                {
+                    // clean up
+                }
+
+                disposed = true;
+
+                base.Dispose(disposing);
+            }
+        }
+
+
+        private void ReadMesh1D()
+        {
+            int count = 0;
+            Invoke(() => UGrid.ug_topology_get_count(fileID, TopologyType.Mesh1dTopology, ref count));
```

Comment:
consider a different name for Invoke. it is also a name of a commonly used c# function